day4_2018.py

```diff
@@ -0,0 +1,94 @@
+with open("input4.txt", "r") as f:
+    inputs = f.read().split("\n")
+    inputs.sort()
+    inputs = [x.split(" ") for x in inputs]
+
+
+def get_guards_ids(inputs):
+    guards_id = []
+    for data_line in inputs:
+        if len(data_line) == 6 and data_line[3][1:] not in guards_id:
+            guards_id.append(data_line[3][1:])
+    return guards_id
+
+
+def get_length_of_next_asleep(idx):
+    guard_asleep_minutes = 0
+    specific_asleep_minutes = []
+    if idx < len(inputs) - 3:
+        wake_up = inputs[idx + 4][1]
+        asslep = inputs[idx + 3][1]
+        if inputs[idx + 3][3] == "asleep":
+            length_of_sleep = int(wake_up[3:5]) - int(asslep[3:5])
+            guard_asleep_minutes += length_of_sleep
+            for minutes in range(int(asslep[3:5]), int(wake_up[3:5])):
+                specific_asleep_minutes.append(minutes)
+    return guard_asleep_minutes, specific_asleep_minutes
+
+
+def get_guard_asleep_minutes(inputs, guard_ids_index):
+    guard_asleep_minutes = []
+    specific_asleep_minutes = []
+    for idx, data_line in enumerate(inputs):
+        if (
+            data_line[3][1:] == gurads_ids[guard_ids_index]
+            and inputs[idx + 1][3] == "asleep"
+        ):
+            wake_up = inputs[idx + 2][1]
+            asslep = inputs[idx + 1][1]
+            length_of_sleep = int(wake_up[3:5]) - int(asslep[3:5])
+            guard_asleep_minutes.append(length_of_sleep)
+            for minutes in range(int(asslep[3:5]), int(wake_up[3:5])):
+                specific_asleep_minutes.append(minutes)
+            if get_length_of_next_asleep(idx)[0] > 0:
+                guard_asleep_minutes.append(get_length_of_next_asleep(idx)[0])
+                guard_asleep_minutes.append(get_length_of_next_asleep(idx + 1)[0])  #
+                specific_asleep_minutes.extend(get_length_of_next_asleep(idx)[1])
```

Comment:
Tu 3 razy wywołujesz funcję od tego samego argumentu `idx`, więc robi się 3x więcej obliczeń niż trzeba. Lepiej byłoby wywołać ją raz, przypisać argumenty i ich później używać, nie wywołując kolejny raz tych samych obliczeń.

Reply:
Tu właśnie zauważyłem że ta linia jest niepotrzebna, bo nie wiedziałem jak to zrobić według Twojego komentarza :D

Reply:
Bardziej mi chodzi konkretnie o `get_length_of_next_asleep(idx)` w liniach 43, 44 i 46. Za każdym razem wywołujesz od nowa tę funkcję.

```suggestion
            zmienna_1, zmienna_2 = get_length_of_next_asleep(idx)
            if zmienna_1 > 0:
                guard_asleep_minutes.append(zmienna_1)
                specific_asleep_minutes.extend(zmienna_2)
```
a nazwy zmiennych to już do Twojej rozkminy

Reply:
W każdym razie jak wywołasz ją raz i przypiszesz do zmiennej, to jest 3x mniej obliczeń w tym miejscu. W tym przypadku pewnie nie zmieni to czasu działania kodu, ale dobrze jest zwracać na to uwagę

Reply:
Prawie tak samo zrobiłem :D Tylko jedną nazwę do zmiennej przypisałem. Nie wiedziałem jak miałbym guard_asleep_minutes.append(get_length_of_next_asleep(idx) też przy takim rozwiązaniu uwzględnić (zakładając że jakiś strażnik mógł spać trzy razy), usunąłem i zadziałało 😅

Reply:
nie do końca kumam, co usunąłeś i co zadziałało? już śmiga to co zjebałeś?

Reply:
Natomiast co do jednej zmiennej to można też tak, ale funkcja zwraca Ci `Tuple`, więc czytelniej jest je [rozpakować](https://www.geeksforgeeks.org/unpacking-a-tuple-in-python/). Takie zmienne są bardziej czytelne niż jakaś tam `zmienna[0]` i `zmienna[1]`. Przy tej drugiej opcji jak ktoś czyta kod to nie wie za bardzo czym jest 0 a czym 1.